operator/watchers/k8s_service_sync.go

```diff
@@ -129,226 +116,73 @@ func StartSynchronizingServices(ctx context.Context, wg *sync.WaitGroup, clients
-			k8sEventMetric(resources.MetricEndpointSlice, resources.MetricUpdate)
-			if oldk8sEP := k8s.ObjToV1EndpointSlice(oldObj); oldk8sEP != nil {
-				if newk8sEP := k8s.ObjToV1EndpointSlice(newObj); newk8sEP != nil {
-					if oldk8sEP.DeepEqual(newk8sEP) {
-						return
-					}
-					K8sSvcCache.UpdateEndpointSlicesV1(newk8sEP, swgEps)
-				}
-			}
-		}
-		delFunc = func(obj interface{}) {
-			k8sEventMetric(resources.MetricEndpointSlice, resources.MetricDelete)
-			k8sEP := k8s.ObjToV1EndpointSlice(obj)
-			if k8sEP == nil {
-				return
-			}
-			K8sSvcCache.DeleteEndpointSlices(k8sEP, swgEps)
-		}
-	} else {
-		esLW = utils.ListerWatcherFromTyped[*slim_discover_v1beta1.EndpointSliceList](slimClient.DiscoveryV1beta1().EndpointSlices(""))
-		objType = &slim_discover_v1beta1.EndpointSlice{}
-		addFunc = func(obj interface{}) {
-			once.Do(func() {
-				// signalize that we have received an endpoint slice
-				// so it means the cluster has endpoint slices enabled.
-				close(hasEndpointSlices)
-			})
-			k8sEventMetric(resources.MetricEndpointSlice, resources.MetricCreate)
-			if k8sEP := k8s.ObjToV1Beta1EndpointSlice(obj); k8sEP != nil {
-				K8sSvcCache.UpdateEndpointSlicesV1Beta1(k8sEP, swgEps)
-			}
-		}
-		updateFunc = func(oldObj, newObj interface{}) {
-			k8sEventMetric(resources.MetricEndpointSlice, resources.MetricUpdate)
-			if oldk8sEP := k8s.ObjToV1Beta1EndpointSlice(oldObj); oldk8sEP != nil {
-				if newk8sEP := k8s.ObjToV1Beta1EndpointSlice(newObj); newk8sEP != nil {
-					if oldk8sEP.DeepEqual(newk8sEP) {
-						return
-					}
-					K8sSvcCache.UpdateEndpointSlicesV1Beta1(newk8sEP, swgEps)
-				}
-			}
-		}
-		delFunc = func(obj interface{}) {
-			k8sEventMetric(resources.MetricEndpointSlice, resources.MetricDelete)
-			k8sEP := k8s.ObjToV1Beta1EndpointSlice(obj)
-			if k8sEP == nil {
-				return
+		for ev := range endpoints.Events(ctx) {
+			switch ev.Kind {
+			case resource.Sync:
+				swg.Done()
+			case resource.Upsert:
+				k8sEventMetric(resources.MetricEndpointSlice, resources.MetricUpdate)
+				K8sSvcCache.UpdateEndpoints(ev.Object, swg)
+			case resource.Delete:
+				k8sEventMetric(resources.MetricEndpointSlice, resources.MetricDelete)
+				K8sSvcCache.DeleteEndpoints(ev.Object.EndpointSliceID, swg)
 			}
-			K8sSvcCache.DeleteEndpointSlices(k8sEP, swgEps)
+			ev.Done(nil)
 		}
-	}
-
-	_, endpointController := informer.NewInformer(
-		esLW,
-		objType,
-		0,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    addFunc,
-			UpdateFunc: updateFunc,
-			DeleteFunc: delFunc,
-		},
-		nil,
-	)
-
-	ctx, cancel := context.WithCancel(ctx)
-	// We use the cancel only if there's no endpoint slice support.
-	// Silence the warnings about possible context leak.
-	var _ = cancel
+	}()
 
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		endpointController.Run(ctx.Done())
+
+		// Wait until service cache is ready. k8sSvcCacheSynced is used by
+		// GetServiceIP() to not query incomplete service cache.
+		swg.Stop()
+		swg.Wait()
```

Comment:
I'm not confident that this change preserves the previous behavior. When the `swg` is stopped, all new calls to `Add` become no-ops. Since the `swg` is immediately stopped here, once a sufficient number of events (based on the number of adds happened before this moment) will have been processed (hence calling `Done()`), `Wait` will unlock, with no guarantee that the list completed (and that all events received previously had actually been processed).

Reply:
It should be correct since we're doing `swg.Add()` and then starting to handle events and doing `swg.Done()` when we receive `resource.Sync`. I'll add a few comments.

Reply:
The issue is that `Add()` is a no-op once the swg is stopped, and since here it is immediately stopped, all adds performed by e.g., `K8sSvcCache.UpdateService` don't have any effect (regardless of whether before or after `resource.Sync`). Differently, all `Done()` executions always decrement the counter (until it reaches zero). Hence, when `K8sSvcCache.UpdateService` is called, Add() would not increment the counter (since the swg is stopped), but `Done()` would decrement it, making it reach 0 before the sync has actually completed.